cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/DeprovisionTask.java

```diff
@@ -33,78 +38,173 @@
-      Transactionals.execute(transactional, dsContext -> {
-        ProvisionerDataset dataset = ProvisionerDataset.get(dsContext, datasetFramework);
-        dataset.putClusterInfo(pollingInfo);
-      });
+  private void execute(ProvisioningTaskInfo taskInfo) throws InterruptedException {
+    ClusterOp op = taskInfo.getClusterOp();
+
+    switch (op.getStatus()) {
+      case REQUESTING_DELETE:
+        deleteCluster(taskInfo);
+        break;
+      case POLLING_DELETE:
+        pollForDelete(taskInfo);
+        break;
+      case FAILED:
+        break;
+      case ORPHANED:
+        // end states
+        break;
+    }
+  }
 
-      ClusterStatus status;
-      while ((status = provisioner.getClusterDetail(provisionerContext,
-                                                    cluster).getStatus()) == ClusterStatus.DELETING) {
-        TimeUnit.SECONDS.sleep(10);
+  private void pollForDelete(ProvisioningTaskInfo taskInfo) throws InterruptedException {
+    try {
+      Cluster cluster = taskInfo.getCluster();
+      while (cluster.getStatus() == ClusterStatus.DELETING) {
+        final Cluster clust = cluster;
+        cluster = Retries.callWithRetries(() -> provisioner.getClusterDetail(provisionerContext, clust),
+                                          retryStrategy, e -> e instanceof RetryableProvisionException);
+        TimeUnit.SECONDS.sleep(2);
       }
 
-      // TODO: CDAP-13246 handle unexpected states and retry
-      switch (status) {
+      switch (cluster.getStatus()) {
         case NOT_EXISTS:
           try {
             provisionerNotifier.deprovisioned(programRunId);
-            Transactionals.execute(transactional, dsContext -> {
-              ProvisionerDataset dataset = ProvisionerDataset.get(dsContext, datasetFramework);
-              dataset.deleteClusterInfo(programRunId);
-            });
           } finally {
             // Delete the keys. We only delete when the cluster is gone.
-            SSHKeyInfo keyInfo = clusterInfo.getSshKeyInfo();
+            SSHKeyInfo keyInfo = taskInfo.getSshKeyInfo();
             if (keyInfo != null) {
               Locations.deleteQuietly(locationFactory.create(keyInfo.getKeyDirectory()), true);
             }
           }
+          ProvisioningTaskInfo doneInfo =
+            new ProvisioningTaskInfo(taskInfo, new ClusterOp(ClusterOp.Type.DEPROVISION, ClusterOp.Status.DELETED),
+                                     cluster);
+          persistClusterInfo(doneInfo);
+          break;
+        case RUNNING:
+          ProvisioningTaskInfo deletingInfo =
+            new ProvisioningTaskInfo(taskInfo,
+                                     new ClusterOp(ClusterOp.Type.DEPROVISION, ClusterOp.Status.REQUESTING_DELETE),
+                                     taskInfo.getCluster());
+          persistClusterInfo(deletingInfo);
+          execute(deletingInfo);
+          break;
+        case CREATING:
```

Comment:
If the cluster is in `CREATING` state, it means the user will have to manually delete the cluster afterwards, right?

Reply:
yeah, anything that is orphaned may require some manual intervention